### PR TITLE
refactor: remove OO patterns

### PR DESCRIPTION
**Pull request by jaschdoc**
After this PR, I'll define a macro that makes the syntax less cumbersome. I'm thinking:

```rust
mist_service!(
    actions:
        "hello": handle_hello,
        "hola": handle_hola
    init:
        do_init
)
```

### CODE REVIEW

File: src/mist.rs [hunk 1]
```diff
@@ -1,64 +1,84 @@
...
-    ) -> Result<&Self, String> {
-        if self.action == action {
-            handler(self.payload.clone(), self.envelope.clone()).map_err(|_| {
-                let mut s = "unable to execute action ".to_owned();
-                s.push_str(action);
-                s
-            })?;
+type ActionHandler = (
+    &'static str,
+    Box<dyn FnOnce(Vec<u8>, Envelope) -> Result<(), String>>,
```
Comment (anbclausen) [inline src/mist.rs:9]: Why Box and dyn?

File: src/mist.rs [hunk 2]
```diff
@@ -1,64 +1,84 @@
...
+    let mut buffer = Vec::new();
+    io::stdin()
+        .read_to_end(&mut buffer)
+        .map_err(|_| "unable to read from stdin")?;
+    Ok(buffer)
 }
 
-pub fn service(args: Vec<&'static str>) -> Result<Mist, String> {
-    Mist::new(args)
+fn get_args() -> Result<(String, Envelope), String> {
```
Comment (anbclausen) [inline src/mist.rs:58]: I don't like sometime having `&'static str` and sometime `String` as the error type. Can we solve it?

File: src/mist.rs [hunk 3]
```diff
@@ -1,64 +1,84 @@
...
+    let args = env::args().rev().take(2);
```
Comment (anbclausen) [inline src/mist.rs:59]: Why two?

File: src/mist.rs [hunk 4]
```diff
@@ -1,64 +1,84 @@
...
+    if args.len() < 3 {
```
Comment (anbclausen) [inline src/mist.rs:60]: Where is this requirement stated? (that we need 3 arguments)?

File: src/mist.rs [hunk 5]
```diff
@@ -1,64 +1,84 @@
...
+        Err("Insufficient program arguments".to_string())
```
Comment (anbclausen) [inline src/mist.rs:61]: Can we just return here? Instead of having an else branch?

File: src/mist.rs [hunk 6]
```diff
@@ -1,64 +1,84 @@
...
+    } else {
+        let mut action = None;
+        let mut envelope = None;
+        for (i, arg) in args.enumerate() {
```
Comment (anbclausen) [inline src/mist.rs:65]: Why not just set `action = args[0]`??

File: src/mist.rs (hunk 4 above)
Comment (jaschdoc) [reply src/mist.rs:60]: It's invoked with `prog_name action envelope` so we need 3 args to access the action and envelope

**Pull request merged**